internal/apex/ctlserver.go

```diff
@@ -0,0 +1,84 @@
+package apex
+
+import (
+	"fmt"
+	"net"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"os"
+)
+
+// TODO make this path configurable
+const UnixSocketPath = "/var/run/apex.sock"
+
+// Assumptions made in here:
+//  - ax.logger is thread-safe
+
+func (ax *Apex) CtlServerStart() error {
+	switch ax.os {
+	case Linux.String():
+		return ax.CtlServerStartLinux()
+
+	case Darwin.String():
+		ax.logger.Info("Ctl interface not yet supported on OSX")
+
+	case Windows.String():
+		ax.logger.Info("Ctl interface not yet supported on Windows")
+	}
+
+	return nil
+}
+
+type ApexCtl struct {
+	ax *Apex
+}
+
+func (ac *ApexCtl) Status(_ string, result *string) error {
+	var statusStr string
+	switch ac.ax.status {
+	case ApexStatusStarting:
+		statusStr = "Starting"
+	case ApexStatusAuth:
+		statusStr = "WaitingForAuth"
+	case ApexStatusRunning:
+		statusStr = "Running"
+	default:
+		statusStr = "Unknown"
+	}
+	res := fmt.Sprintf("Status: %s\n", statusStr)
+	if len(ac.ax.statusMsg) > 0 {
+		res += ac.ax.statusMsg
+	}
+	*result = res
+	return nil
+}
+
+func (ax *Apex) CtlServerStartLinux() error {
+	os.Remove(UnixSocketPath)
+	l, err := net.ListenUnix("unix", &net.UnixAddr{Name: UnixSocketPath, Net: "unix"})
+	if err != nil {
+		ax.logger.Error("Error creating unix socket: ", err)
+		return err
+	}
+	go func() {
+		defer l.Close()
+
+		ac := new(ApexCtl)
+		ac.ax = ax
+		err := rpc.Register(ac)
+		if err != nil {
+			ax.logger.Errorln("Error on rpc.Register(): ", err)
+		}
+
+		for {
+			conn, err := l.Accept()
+			if err != nil {
+				ax.logger.Error("Error on Accept(): ", err)
+				continue
+			}
+			go jsonrpc.ServeConn(conn)
```

Comment:
We should probably use a `sync.WaitGroup` ourselves here.
So:
```go
var wg sync.WaitGroup

select {
    case <-stop:
      // handle shutdown
      // wait for existing rpcs to finish
      wg.Wait()
      return
    default:
      conn, err := l.Accept()
      if err != nil {
	ax.logger.Error("Error on Accept(): ", err)
	continue
     }
     wg.Add(1)
     go jsonrpc.ServeConn(conn)

  }
}
```

Reply:
done